### ISSUE TITLE
Codegen: float type and float functions in c

### ISSUE THREAD

**Issue by stewpend0us** [sympy/sympy#14063]
To get codegen to use float types AND float functions in c I had to do the following:

```python
from sympy.utilities.codegen import codegen, get_code_generator, default_datatypes
from sympy.codegen.ast import real, float32

cg = get_code_generator('c', project='example', standard='c99') #note project must be specified here
cg.printer.type_aliases[real] = float32 #makes the functions correct (cosf instead of cos)
default_datatypes["float"].cname = "float" #makes the data-type correct (float instead of double)

codegen(
       	functions,
	prefix="demo",
       	to_files=True,
	code_gen=cg
       	)
```

Is this the "right" way go about this? Could this be an option on codegen to save someone else the pain i the future?

Unrelated rant:
No offence to anyone but I found the code to be very convoluted and hard to follow. Figuring this out took me several hours...more likely a personal problem but whew that was annoying.

**Comment by moorepants** [comment]
This is brand new functionality I think, from @bjodah's work this past summer. It may not be documented.

**Comment by asmeurer** [comment]
> No offence to anyone but I found the code to be very convoluted and hard to follow. 

Are you referring to sympy.utilities.codegen or sympy.codegen? If the former, then I completely agree. I would prefer to completely replace it with better alternatives, so that it isn't an issue any more. Some cleanups to it would be nice as well (though there's a limit to how much we can do without breaking the API). 

If you mean sympy.codegen, then that's a more important issue, because that code is still very new and in active development. Any issues there we should try to clean up now while the API is still at least somewhat provisional. 

By the way, you may be interested in https://github.com/sympy/sympy/pull/13100.

**Comment by stewpend0us** [comment]
Ah. Yes I'm definitely working with sympy.utilities.codegen I didn't realize sympy.codegen was a thing. Perhaps that would have been a better place to start?

Forgive me I'm fairly new to github/collaboration #13100 looks like a codegen PR that's in progress? Would that PR be a good place to contribute or would I need to make my own PR?

**Comment by bjodah** [comment]
I'm fine with you working directly against against that branch. You can do like this (in your sympy repo):
```
git remote add bjodah git://github.com/bjodah
git fetch bjodah
git checkout codegen-ast-more-nodes
# edit files..
git commit -am "commit message goes here"
git push --set-upstream origin codegen-ast-more-nodes
```
then you can open a pull request from your branch to my branch if you'd like.

**Comment by asmeurer** [comment]
> Ah. Yes I'm definitely working with sympy.utilities.codegen I didn't realize sympy.codegen was a thing. Perhaps that would have been a better place to start?

sympy.codegen is still very new, so I would recommend checking it out, but don't be surprised if something you want isn't implemented yet or if you find a bug (in either case, please open an issue). The basic idea behind it is to construct the thing you want to code generate as an AST, then use the code printers (CCodePrinter, ccode, etc.) to print them directly.  

And by the way you are using sympy.codegen in your example (that's why I asked).

**Comment by asmeurer** [comment]
Also as I noted the API for sympy.codegen is still provisional, so if you have any suggestions on how to improve it, that's still possible (once it goes in a release it's a lot harder to change the API, so we'd like to fix any bumps now if possible).

**Comment by stewpend0us** [comment]
so even though i'm importing from sympy.utilities.codegen I'm using sympy.codegen? I did have to mess around with the code printer stuff to get the job done...but I didn't realize I was using the new stuff.

Thanks for the help!

**Comment by bjodah** [comment]
I was actually surprised that using the `float32` class from `sympy.codegen` worked (somewhat) together with the old `sympy.utilities.codegen`. (it worked because both rely on the printers in `sympy.printing`).

And as Aaron said, we really value all feedback (on `sympy.codegen` in particular since we can change it more freely still).

**Comment by stewpend0us** [comment]
Yeah it was strange how half of it needed settings on the printer to be changed and the other part was a local dictionary.

Derp I didn't realize the real and float32 were coming from sympy.codegen ... what was I doing?? haha

I'm about to take a look at the new stuff :)

**Comment by stewpend0us** [comment]
I've been working with codegen a bit more ( actually on your branch @bjodah ). What's the best way to give feedback/ask questions?

**Comment by bjodah** [comment]
What about you opening a PR against my branch like the last time? Do you want me to merge https://github.com/bjodah/sympy/pull/2 before?

**Comment by stewpend0us** [comment]
I don't really have any code to do a pull request with...more just questions about what's going on? I guess I could open an issue on your repo

**Comment by bjodah** [comment]
Sure, that works for me.

**Comment by stewpend0us** [comment]
oh..doesn't look like you have issues enabled or something?
